azure/services/scalesets/client.go

```diff
@@ -97,166 +104,128 @@ func (ac *AzureClient) ListInstances(ctx context.Context, resourceGroupName stri
+	scaleset, ok := parameters.(armcompute.VirtualMachineScaleSet)
+	if !ok && parameters != nil {
+		return nil, nil, errors.Errorf("%T is not an armcompute.VirtualMachineScaleSet", parameters)
 	}
 
-	createFuture, err := ac.scalesets.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.ResourceName(), scaleset)
+	opts := &armcompute.VirtualMachineScaleSetsClientBeginCreateOrUpdateOptions{ResumeToken: resumeToken}
+	poller, err = ac.scalesets.BeginCreateOrUpdate(ctx, spec.ResourceGroupName(), spec.ResourceName(), scaleset, opts)
 	if err != nil {
 		return nil, nil, err
 	}
 
 	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
 	defer cancel()
 
-	err = createFuture.WaitForCompletionRef(ctx, ac.scalesets.Client)
+	pollOpts := &runtime.PollUntilDoneOptions{Frequency: asyncpoller.DefaultPollerFrequency}
+	resp, err := poller.PollUntilDone(ctx, pollOpts)
 	if err != nil {
-		// if an error occurs, return the future.
+		// if an error occurs, return the poller.
 		// this means the long-running operation didn't finish in the specified timeout.
-		return nil, &createFuture, err
+		return nil, poller, err
 	}
 
-	result, err = createFuture.Result(ac.scalesets)
-	// if the operation completed, return a nil future
-	return result, nil, err
+	// if the operation completed, return a nil poller
+	return resp.VirtualMachineScaleSet, nil, err
 }
 
 // UpdateInstances update instances of a VM scale set.
-func (ac *AzureClient) UpdateInstances(ctx context.Context, resourceGroupName, vmssName string, instanceIDs []string) error {
+func (ac *AzureClient) UpdateInstances(ctx context.Context, resourceGroupName, vmssName string, instanceIDs []string, resumeToken string) error {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "scalesets.AzureClient.UpdateInstances")
 	defer done()
 
-	params := compute.VirtualMachineScaleSetVMInstanceRequiredIDs{
-		InstanceIds: &instanceIDs,
+	params := armcompute.VirtualMachineScaleSetVMInstanceRequiredIDs{
+		InstanceIDs: azure.PtrSlice(&instanceIDs),
 	}
-	future, err := ac.scalesets.UpdateInstances(ctx, resourceGroupName, vmssName, params)
+	opts := &armcompute.VirtualMachineScaleSetsClientBeginUpdateInstancesOptions{ResumeToken: resumeToken}
+	poller, err := ac.scalesets.BeginUpdateInstances(ctx, resourceGroupName, vmssName, params, opts)
 	if err != nil {
 		return err
 	}
-	err = future.WaitForCompletionRef(ctx, ac.scalesets.Client)
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
+	defer cancel()
+
+	pollOpts := &runtime.PollUntilDoneOptions{Frequency: asyncpoller.DefaultPollerFrequency}
+	_, err = poller.PollUntilDone(ctx, pollOpts)
 	if err != nil {
+		// if an error occurs, return the poller.
+		// this means the long-running operation didn't finish in the specified timeout.
 		return err
 	}
-	_, err = future.Result(ac.scalesets)
+
+	// TODO: Retry this operation internally to this function so it appears synchronous
```

Comment:
/hold

Still need to figure this out. It's an async SDK function now, but the `asyncpoller` framework only handles CreateOrUpdate and Delete.

Reply:
Is this even being used at all? A plain grep for `UpdateInstances` on main only really shows this definition and the mock.

Reply:
Perhaps it isn't used: the relevant e2e tests passed even though the current implementation is probably broken. I'll see about removing this code.

Reply:
/hold cancel

Yep, `UpdateInstances` was unused and is now gone. Thanks @nojnhuh! I think this is ready for review now.